[gohyongjing] iP

gohyongjing opened the pull request:
# Duke

> “Your mind is for *having* ideas, not *holding* them.” – David Allen ([source](https://dansilvestre.com/productivity-quotes))

Added some improvements to **Duke**:

* Storage support to save tasks to the disk
* OOP for abstraction
* Automatic build using gradle

To get started,

1. Download it from [here](https://github.com/gohyongjing/ip/releases/tag/v0.1).
1. Move the jar file into an empty folder
1. Open a command window in that folder
1. Run the command java -jar Duke.jar
1. Let it manage your tasks for you :wink:

Completed features:
- [x] ~~Level-1~~
- [x] ~~Level-2~~
- [x] ~~Level-3~~
- [x] ~~Level-4~~
- [x] ~~Level-5~~
- [x] ~~Level-6~~
- [x] ~~Level-7~~
- [x] ~~Level-8~~
- [x] ~~Level-9~~
- [ ] Level-10

Heres the `main` method:
```java
/**
 * Entry point to start and run duke.
 *
 * @param args Command line arguments.
 */
public static void main(String[] args) {
    Duke duke = new Duke();
    duke.run();
}
```

Commit:
Add Gradle support

gohyongjing committed:
Add skeletal program for Duke

gohyongjing committed:
Add feature to store tasks

gohyongjing committed:
Refactor task into a separate class

gohyongjing committed:
Add feature to mark tasks as done

gohyongjing committed:
Add subclasses for tasks

gohyongjing committed:
Add support for different tasks

gohyongjing committed:
Add test case

gohyongjing committed:
Refactor command into a class

gohyongjing committed:
Add error handling

gohyongjing committed:
Refactor Duke to use ArrayList of Tasks

gohyongjing committed:
Refactor Command to use ArrayList of Strings

gohyongjing committed:
Add delete task feature

gohyongjing committed:
Add test cases for error handling

gohyongjing committed:
Add enums for Command Arguments

gohyongjing committed:
Clean up code and write documentation

gohyongjing committed:
Move classes into packages

gohyongjing committed:
Remove unnecessary 'this' keywords

gohyongjing committed:
Update javadocs to follow convention

gohyongjing committed:
Refactor tasklist into a class

tasks are currently stored in a list in the duke class.

Refactoring tasks list into a seperate file makes the duke class less cluttered when implementing persistent storage later.

Let's refactor the ArrayList into a separate class to make further modifications easier.

gohyongjing committed:
Refactor TaskList to further handle commands

gohyongjing committed:
Add persistent storage of tasks

gohyongjing committed:
Add date validation adn formatting

gohyongjing committed:
Add documentation and reorganise saved data

gohyongjing committed:
Merge branch 'branch-Level-7'

gohyongjing committed:
Resolve merge conflicts

gohyongjing committed:
Add Storage class to handle file i/o

gohyongjing committed:
Add Parser class to handle parsing of commands

gohyongjing committed:
Add Ui class to handle user interface

gohyongjing committed:
Refactor duke into a package

gohyongjing committed:
Merge branch 'add-gradle-support'

gohyongjing committed:
Add gradle to build and run duke

gohyongjing committed:
Set up basic unit testing

gohyongjing committed:
Add Deadline test case

gohyongjing committed:
Add javadocs

gohyongjing committed:
Set up checkstyle with gradle

gohyongjing committed:
Fix styles acoording to coding standard

gohyongjing committed:
Add find keyword feature

gohyongjing committed:
Merge branch 'branch-A-CodingStandard'

gohyongjing committed:
Resolve merge conflicts

gohyongjing committed:
Add skeletal files for GUI

gohyongjing committed:
Set up views

gohyongjing committed:
Add GUI

gohyongjing committed:
Fix datetime bug affecting storage parsing

gohyongjing committed:
Update Deadline Test

gohyongjing committed:
Remove obsolete command line tests

gohyongjing committed:
Merge branch 'branch-Level-10'

gohyongjing committed:
Improve user experience.

GUI only shows the first 5 tasks on command 'list' and abruptly closes the window on commnd 'bye'.

This provides the user bad with a bad experience.

Let's allow the dialogue box to expand as needed and put a 5 second delay after 'bye' command is entered to allow users to read the response properly.

gohyongjing committed:
Merge branch 'branch-Level-10'

zm-l commented on src/main/java/duke/command/Parser.java:
```diff
@@ -0,0 +1,112 @@
+        LinkedHashMap<Command.Argument, String> arguments = new LinkedHashMap<>();
+        for (String term : input.strip().split(" /")) {
+            int firstSpace = term.indexOf(" ");
+            Command.Argument argument = Command.parseArgument((firstSpace == -1
```
I am not too sure about this but 
···
alpha = (aLongBooleanExpression) ? beta : gamma;
alpha = (aLongBooleanExpression)
        ? beta
        : gamma;
···
so should the Boolean be bracketed?

zm-l replied:
I am not too sure about this but
···
alpha = (aLongBooleanExpression) ? beta : gamma;
alpha = (aLongBooleanExpression)
? beta
: gamma;
···
so should the Boolean be bracketed?

zm-l replied:
And will it be better to put the ternary operator on a new line?

zm-l commented on src/main/java/duke/command/Parser.java:
```diff
@@ -0,0 +1,112 @@
+     *
+     * @param command Command to check.
+     * @param args Required arguments.
+     * @throws IllegalArgumentException if any of the required arguments are not in
```
Should the "if" be capitalized?

zm-l commented on src/main/java/duke/task/TaskList.java:
```diff
@@ -0,0 +1,149 @@
+     */
+    public Task execute(Command command) {
+        Task task = null;
+        switch (command.getName()) {
```
I like how you use switch to make the code looks clean.

zm-l commented on src/main/java/duke/ui/MainWindow.java:
```diff
@@ -0,0 +1,186 @@
+import javafx.scene.image.Image;
+import javafx.scene.layout.AnchorPane;
+import javafx.scene.layout.VBox;
+/**
```
Might want to add a new line for the sake of consistency with other class.

zm-l reviewed (commented):
LGTM.
Overall, your code abides the coding standards perfectly and there are sufficient comments for ease of understanding your code.

gohyongjing commented:
> LGTM. Overall, your code abides the coding standards perfectly and there are sufficient comments for ease of understanding your code.

Thanks! I will take note and make the changes as suggested

gohyongjing committed:
Update build.gradle with new entry point

build.gradle currently points to Duke.java for its entry point.

With JavaFX, we will need a separate launcher class to run the app.

Let's update the entry point to point to the launcher class instead.

gohyongjing committed:
Update build.gradle with new entry point

gohyongjing committed:
Merge branch 'master' of github.com:gohyongjing/ip

gohyongjing committed:
Add assertions to verify internal state of duke.

No assertions are currently in place.

This is a potential source of bugs if we add a new command and forget to handle it as it fails silently. If fxml is changed or image is removed, some objects might be null.

Let's introduce some assertions to detect these sources of error.

gohyongjing committed:
Improve code quality based on review comments.

Some javadoc comments are written without capitalisation, some new lines are missing and some ternary operators are too long.

This makes the code inconsistent and hard to read.

Let's fix these issues according to the SE-EDU coding quality recommendations.
* add a new line after the import before the class comment.
* capitalise all @throws description in javadocs.
* rearrange parenthesis for long ternary operators.

gohyongjing committed:
Set up continuous integration

gohyongjing committed:
Merge pull request #1 from gohyongjing/branch-A-Assertions

Add assertions to verify internal state of duke.

gohyongjing committed:
Resolve merge conflicts

gohyongjing committed:
Merge pull request #2 from gohyongjing/branch-A-CodeQuality

Improve code quality based on review comments.

gohyongjing committed:
Change date format.

DateTime class uses localised date.

This causes the format to be inconsistent in different geological areas.

Let's change the format to one that is constistent accross geographical locations.

gohyongjing committed:
Merge pull request #3 from gohyongjing/branch-A-CI

Set up continuous integration

gohyongjing committed:
Update tasklist to reject duplicate tasks

gohyongjing committed:
Resolve merge conflicts

gohyongjing committed:
Merge pull request #4 from gohyongjing/branch-C-DetectDuplicates

Update tasklist to reject duplicate tasks

gohyongjing committed:
Add more tests for tasks and subclasses

gohyongjing committed:
Handle invalid index for mark and delete

gohyongjing committed:
Fix bug in mark and delete causing nullPointerException

gohyongjing committed:
Merge pull request #6 from gohyongjing/branch-A-MoreErrorHandling

Branch a more error handling

gohyongjing committed:
Fix bug causing disk write failure when directory exists

gohyongjing committed:
Merge branch 'master' of github.com:gohyongjing/ip

gohyongjing committed:
Update user guide

gohyongjing committed:
Fix bug causing task list to be truncated in user interface.

DialogBox uses Labels for text.

This causes the task list to be truncated for long outputs of list and ind commands.

Let's replace Label with Text component to prevent this truncation.

gohyongjing committed:
Fix bug causing user icon to move out of window

gohyongjing committed:
Fix merge conflict

gohyongjing committed:
Merge pull request #7 from gohyongjing/branch-A-UserGuide

Update user guide

gohyongjing committed:
Fix bug causing user icon to be cut off due to no text wrap

gohyongjing committed:
Fix merge conflict

gohyongjing committed:
Handle non integer argument for mark and delete

gohyongjing committed:
Add javadoc for TaskList.parseIndex

gohyongjing committed:
Fix javadoc for datetime
